### PR TITLE
Add option to ignore_values for dict_to_protobuf

### PR DESCRIPTION
**Pull request by pfbrousseau**
Brought up by FAIR-8172
Resolves problem where default enum would need to be an empty string as passed by another service: ```KeyError: '`` is not a valid value for field [enum name]'```

### CODE REVIEW

File: protobuf_to_dict/convertor.py [hunk 1]
```diff
@@ -192,25 +195,33 @@ def _get_field_mapping(pb, dict_value, strict):
...
-def _dict_to_protobuf(pb, value, type_callable_map, strict, ignore_none):
+def _dict_to_protobuf(pb, value, type_callable_map, strict, ignore_none, ignore_values):
     fields = _get_field_mapping(pb, value, strict)
 
     for field, input_value, pb_value in fields:
-        if ignore_none and input_value is None:
+        if ignore_none and input_value is None:  # To deprecate. Use `ignore_values={None}` instead.
             continue
+        if ignore_values:
+            try:
```
Comment (seperman) [inline protobuf_to_dict/convertor.py:205]: As you said, it looks kind of ugly to have the try except here.
What do you think about:

`if getattr(input_value, '__hash__') and input_value in ignore_values`,
then you won't need the try except clause.

File: protobuf_to_dict/convertor.py [hunk 2]
```diff
@@ -192,25 +195,33 @@ def _get_field_mapping(pb, dict_value, strict):
     return field_mapping
 
 
-def _dict_to_protobuf(pb, value, type_callable_map, strict, ignore_none):
+def _dict_to_protobuf(pb, value, type_callable_map, strict, ignore_none, ignore_values):
```
Comment (seperman) [inline protobuf_to_dict/convertor.py:198]: What do you think about converting ignore_values into set if it is not already a set. That way if somebody passed a list, we still use the set for contain operation.

File: tests/test_proto_to_dict.py [hunk 3]
```diff
@@ -164,6 +164,38 @@ def test_nested_ignore_none(self):
         m2 = dict_to_protobuf(MessageOfTypes, d, ignore_none=True)
         assert m2.nestedMap['123'].req == ''
 
+    def test_ignore_values_as_ignore_none(self):
```
Comment (seperman) [inline tests/test_proto_to_dict.py:167]: Did you want to test it specifically with enums and empty strings too?

**Review by seperman — COMMENTED** [review]
Hi @pfbrousseau 
Thanks for taking the time to making protobuf-to-dict better. I left a few comments.

**Comment by mvillalobosj** [comment]
I feel like instead of adding this kind of blanket "ignore_values" parameter which adds another check to the for loop and doesn't allow for fine-grain control (i.e. I want to ignore a value for _this_ field but not _that_ field), this could have just been an extension of how the `strict` parameter in the `_string_to_enum` method is currently implemented to solve your original problem.

https://github.com/wearefair/protobuf-to-dict/blob/master/protobuf_to_dict/convertor.py#L244

I think changing it from this:
```python
def _string_to_enum(field, input_value, strict=False):
    try:
        input_value = field.enum_type.values_by_name[input_value].number
    except KeyError:
        if strict:
            raise KeyError("`%s` is not a valid value for field `%s`" % (input_value, field.name))
        else:
            return _string_to_enum(field, input_value.upper(), strict=True)
    return input_value
```
to this:
```python
def _string_to_enum(field, input_value, strict=False):
    try:
        return field.enum_type.values_by_name[input_value].number
    except KeyError:
        if strict:
            raise KeyError("`%s` is not a valid value for field `%s`" % (input_value, field.name))
        if not input_value:
            return 0
        return _string_to_enum(field, input_value.upper(), strict=True)  
```

or you could even argue that if `strict` is False and we don't recognize the enum value, we should just return 0 and not set it (although this would probably break current functionality... maybe test the uppercase too by default, as this is a common case).

```python
def _string_to_enum(field, input_value, strict=False):
    try:
        return field.enum_type.values_by_name[input_value].number
    except KeyError:
        try:
            return field.enum_type.values_by_name[input_value.upper()].number
        except KeyError as err:
            if strict:
                raise KeyError("`%s` is not a valid value for field `%s`" % (input_value, field.name)) from err
    return 0
```

Would be a much more targeted approach towards solving the specific issue you are having. I worry that having the `ignore_values` gives you too much control over the underlying algorithm (but not the ability to focus that control) and it will enable you to more easily pursue anti-patterns with the protobuf definitions, your api or data model. Whereas if you have those anti-patterns now, you have to go to further pains (i.e. massaging the incoming data) in order to get it to fit with this algorithm which should be a red-flag that you are doing something wrong.

**Comment by pfbrousseau** [comment]
Fixing ignore_null and fixing my data source took care of this.

**Pull request closed without merging**